h2o-core/src/test/java/water/ExternalFrameWriterClientTest.java

```diff
@@ -27,97 +28,60 @@ public static void setup() {
-                        }
-                        writer.sendInt(0);
-                        writer.sendBoolean(true);
-                        writer.sendString(null);
-                        writer.sendTimestamp(time);
-
-                        writer.sendInt(1);
-                        writer.sendBoolean(true);
-                        writer.sendString("\u0080");
-                        writer.sendTimestamp(time);
-
-                        // send NA for all columns
-                        writer.sendNA();
-                        writer.sendNA();
-                        writer.sendNA();
-                        writer.sendNA();
+                writer.sendInt(1);
+                writer.sendBoolean(true);
+                writer.sendString("\u0080");
+                writer.sendTimestamp(time);
 
-                        try {
-                            writer.waitUntilAllWritten(10);
-                        }catch (ExternalFrameConfirmationException e){
-                            e.printStackTrace();
-                        }
-                        sock.close();
-                        rowsPerChunk[currentIndex] = 1000;
-                    } catch (IOException ignore) {
-                    }
-                }
+                // send NA for all columns
+                writer.sendNA();
+                writer.sendNA();
+                writer.sendNA();
+                writer.sendNA();
+            }
 
-            };
-            threads[idx].start();
-        }
+            @Override
+            public int nrows() {
+                return 1000;
+            }
 
-        // wait for all writer thread to finish
-        for (Thread t : threads) {
-            try {
-                t.join();
-            } catch (InterruptedException e) {
-                e.printStackTrace();
+            @Override
+            public String[] colNames() {
+                return new String[] {"NUM", "BOOL", "STR", "TIMESTAMP"};
             }
-        }
 
+            @Override
+            public byte[] colTypes() {
+                return new byte[] {
+                    ExternalFrameUtils.EXPECTED_INT,
+                    ExternalFrameUtils.EXPECTED_BOOL,
+                    ExternalFrameUtils.EXPECTED_STRING,
+                    ExternalFrameUtils.EXPECTED_TIMESTAMP};
+            }
+        };
 
-        ChunkUtils.finalizeFrame(frameName, rowsPerChunk, ExternalFrameUtils.vecTypesFromExpectedTypes(expectedTypes, null), null);
+        final String[] nodes = getH2ONodes();
+        // we will open 2 connection per h2o node
+        final String[] connStrings = ArrayUtils.join(nodes, nodes);
 
-        Frame frame = null;
+        Frame frame = createFrame(testOp, connStrings);
         try {
-            frame = DKV.getGet(frameName);
             assertEquals(frame.anyVec().nChunks(), connStrings.length);
             assertEquals(frame._names.length, 4);
             assertEquals(frame.numCols(), 4);
```

Comment:
I don't like that it is only checking the values that are filled. Ideally we should check the whole frame.

Also the comment "// last row should be NA" is misplaced, it should ideally be on line 104 of the original source file.